### PR TITLE
JAVA-1256: Log parameters of BuiltStatement in QueryLogger

### PR DESCRIPTION
**Pull request by adutra**

### CODE REVIEW

**Review by tolbertam — APPROVED** [review]
looks great :+1:

File: driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java [hunk 1]
```diff
@@ -865,6 +868,38 @@ protected String parameterValueAsString(Object value) {
     return valueStr;
   }
 
+  protected int appendParameters(BuiltStatement statement, StringBuilder buffer, int remaining) {
+    if (remaining == 0) {
+      return 0;
+    }
+    int numberOfParameters =
+        statement.getValues(protocolVersion(), cluster.getConfiguration().getCodecRegistry())
```
Comment (codefromthecrypt) [inline driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java:876]: Hi. I noticed this causes an NPE for scenarios like this as getValues can return null.

```java
      session().execute(QueryBuilder.select("trace_id").from("span").limit(1));
```

I'm not saying this is the only path, but after substantial debugging, narrowed down to this commit as the commit prior passes and when enabling debug on logger on `Connection`. The user behavior though not knowing about this logger is pretty drastic as the connection basically hangs until you turn off query logging. This is likely also a problem in 4.x as when trying to migrate from there it led me to try on our current library which is using 3.x and had the same hang.

```
2020-08-02 09:54:45:211 [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection - Connection[/127.0.0.1:9042-2, inFlight=0, closed=false] connection error
java.lang.NullPointerException
	at com.datastax.driver.core.QueryLogger.appendParameters(QueryLogger.java:876)
	at com.datastax.driver.core.QueryLogger.logQuery(QueryLogger.java:709)
	at com.datastax.driver.core.QueryLogger.maybeLogNormalQuery(QueryLogger.java:671)
	at com.datastax.driver.core.QueryLogger$ConstantThresholdQueryLogger.maybeLogNormalOrSlowQuery(QueryLogger.java:294)
	at com.datastax.driver.core.QueryLogger.update(QueryLogger.java:654)
	at com.datastax.driver.core.Cluster$Manager.reportQuery(Cluster.java:1881)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:829)
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1287)
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1205)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at com.datastax.driver.core.InboundTrafficMeter.channelRead(InboundTrafficMeter.java:38)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

Note: I tried to raise a JIRA on this but the email linked to my github, and what I use in the mailing list, which is what I prefer use, says it is revoked for doing so. Hence, pinging the issue related directly and hoping folks are watching.

Comment (codefromthecrypt) [reply driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java:876]: incidentally, this bug happens at TRACE level of the query logger (despite the error showing here in the other logger at debug)

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java:876]: Hi, indeed I can reproduce the bug very easily, so we are going to provide a fix for this soon. Thanks for reporting!

> Note: I tried to raise a JIRA on this but the email linked to my github, and what I use in the mailing list, which is what I prefer use, says it is revoked for doing so.

I didn't fully understand that part, but anyways I created a Jira already: [JAVA-2857](https://datastax-oss.atlassian.net/browse/JAVA-2857). If your problem with our Jira persists could you please share your Jira login with me and I will check.

> This is likely also a problem in 4.x as when trying to migrate from there it led me to try on our current library which is using 3.x and had the same hang.

Could you please provide me the stack trace for 4.x? Looking at `com.datastax.oss.driver.internal.core.tracker.RequestLogFormatter` I don't see how we could have a NPE there. I also think that any errors bubbling up from a `com.datastax.oss.driver.internal.core.tracker.RequestLogger#onSuccess` call will be correctly intercepted in 4,x – but I might be wrong.

> incidentally, this bug happens at TRACE level of the query logger (despite the error showing here in the other logger at debug)

Yes it's a bit twisted, the way it works is that you need to activate TRACE to get the parameters, but everything is logged at DEBUG.

**Comment by codefromthecrypt** [comment]
Thanks for the help!

> Could you please provide me the stack trace for 4.x? Looking at
> com.datastax.oss.driver.internal.core.tracker.RequestLogFormatter I don't
> see how we could have a NPE there. I also think that any errors bubbling up
> from a
> com.datastax.oss.driver.internal.core.tracker.RequestLogger#onSuccess
> call will be correctly intercepted in 4,x – but I might be wrong.
>

4.x could well work. I need to retest. I stopped my mess of a 4.x upgrade
when I struggled with the query logging then noticed this problem. There
could well be no relationship, so you can consider this a non-issue in 4.x
unless I report back contrary

File: driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java (hunk 1 above)
Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java:876]: PR is up here FYI: https://github.com/datastax/java-driver/pull/1488.

**Pull request merged**